File: study_tracker_flutter/lib/core/utils/analytics_utils.dart
import '../../domain/domain.dart';

// ─────────────────────────────────────────────────────────────────────────────
// Analytics computation (ported from analyticsUtils.ts)
// ─────────────────────────────────────────────────────────────────────────────

import 'utils.dart';

class DailyTotal {
  final String date; // YYYY-MM-DD
  final int minutes;
  const DailyTotal({required this.date, required this.minutes});
}

class WeeklyTotal {
  final String weekStart;
  final int minutes;
  final int sessionCount;
  const WeeklyTotal({
    required this.weekStart,
    required this.minutes,
    required this.sessionCount,
  });
}

class SubjectBreakdown {
  final String name;
  final int minutes;
  final int sessionCount;
  final int percentage;
  const SubjectBreakdown({
    required this.name,
    required this.minutes,
    required this.sessionCount,
    required this.percentage,
  });
}

class HourlyDistribution {
  final int hour;
  final int minutes;
  const HourlyDistribution({required this.hour, required this.minutes});
}

class MoodDistribution {
  final String mood;
  final int count;
  final int percentage;
  const MoodDistribution({
    required this.mood,
    required this.count,
    required this.percentage,
  });
}

class StudyBreakRatio {
  final int studyMinutes;
  final int breakMinutes;
  final String ratio;
  const StudyBreakRatio({
    required this.studyMinutes,
    required this.breakMinutes,
    required this.ratio,
  });
}

/// Compute daily totals filling in zero-days for the full range.
List<DailyTotal> computeDailyTotals(
  List<StudySession> sessions,
  DateTime startDate,
  DateTime endDate,
) {
  final map = <String, int>{};
  for (final d in dateRange(startDate, endDate)) {
    final key = '${d.year}-${d.month.toString().padLeft(2, '0')}-${d.day.toString().padLeft(2, '0')}';
    map[key] = 0;
  }
  for (final s in sessions) {
    final d = s.startAt;
    final key = '${d.year}-${d.month.toString().padLeft(2, '0')}-${d.day.toString().padLeft(2, '0')}';
    if (map.containsKey(key)) {
      map[key] = map[key]! + s.durationMinutes;
    }
  }
  return map.entries
      .map((e) => DailyTotal(date: e.key, minutes: e.value))
      .toList()
    ..sort((a, b) => a.date.compareTo(b.date));
}

/// Compute weekly totals for N weeks back.
List<WeeklyTotal> computeWeeklyTotals(
    List<StudySession> sessions, int weeksBack) {
  final now = DateTime.now();
  final result = <WeeklyTotal>[];

  for (int i = weeksBack - 1; i >= 0; i--) {
    final ws = startOfWeek(now.subtract(Duration(days: i * 7)));
    final we = ws.add(const Duration(days: 7)).subtract(const Duration(milliseconds: 1));
    final filtered = sessions
        .where((s) => !s.startAt.isBefore(ws) && !s.startAt.isAfter(we))
        .toList();
    result.add(WeeklyTotal(
      weekStart:
          '${ws.year}-${ws.month.toString().padLeft(2, '0')}-${ws.day.toString().padLeft(2, '0')}',
      minutes: filtered.fold(0, (sum, s) => sum + s.durationMinutes),
      sessionCount: filtered.length,
    ));
  }
  return result;
}

/// Compute subject breakdown with percentages.
List<SubjectBreakdown> computeSubjectBreakdown(List<StudySession> sessions) {
  final map = <String, (int minutes, int count)>{};
  int total = 0;
  for (final s in sessions) {
    final name = s.subjectName ?? 'Uncategorized';
    final prev = map[name] ?? (0, 0);
    map[name] = (prev.$1 + s.durationMinutes, prev.$2 + 1);
    total += s.durationMinutes;
  }
  return map.entries
      .map((e) => SubjectBreakdown(
            name: e.key,
            minutes: e.value.$1,
            sessionCount: e.value.$2,
            percentage:
                total > 0 ? ((e.value.$1 / total) * 100).round() : 0,
          ))
      .toList()
    ..sort((a, b) => b.minutes.compareTo(a.minutes));
}

/// Compute 24-hour distribution.
List<HourlyDistribution> computeHourlyDistribution(
    List<StudySession> sessions) {
  final hours = List<int>.filled(24, 0);
  for (final s in sessions) {
    hours[s.startAt.hour] += s.durationMinutes;
  }
  return List.generate(
      24, (i) => HourlyDistribution(hour: i, minutes: hours[i]));
}

/// Compute mood distribution.
List<MoodDistribution> computeMoodDistribution(List<StudySession> sessions) {
  final map = <String, int>{};
  int total = 0;
  for (final s in sessions) {
    if (s.mood != null && s.mood!.isNotEmpty) {
      map[s.mood!] = (map[s.mood!] ?? 0) + 1;
      total++;
    }
  }
  return map.entries
      .map((e) => MoodDistribution(
            mood: e.key,
            count: e.value,
            percentage: total > 0 ? ((e.value / total) * 100).round() : 0,
          ))
      .toList()
    ..sort((a, b) => b.count.compareTo(a.count));
}

/// Study/break ratio.
StudyBreakRatio computeStudyBreakRatio(List<StudySession> sessions) {
  final study = sessions.fold(0, (s, e) => s + e.durationMinutes);
  final brk = sessions.fold(0, (s, e) => s + e.breakMinutes);
  final ratio = brk > 0
      ? '${(study / brk).toStringAsFixed(1)}:1'
      : study > 0
          ? '∞:1'
          : '0:0';
  return StudyBreakRatio(
      studyMinutes: study, breakMinutes: brk, ratio: ratio);
}

/// Rolling average.
int computeRollingAverage(List<DailyTotal> dailyTotals, int windowDays) {
  if (dailyTotals.isEmpty) return 0;
  final recent = dailyTotals.length > windowDays
      ? dailyTotals.sublist(dailyTotals.length - windowDays)
      : dailyTotals;
  final total = recent.fold(0, (s, d) => s + d.minutes);
  return (total / recent.length).round();
}

File: src/core/utils/analyticsUtils.ts
/** Analytics computation utilities */

import type { StudySession } from '../domain';
import { startOfDay, startOfWeek, dateRange, getHour } from './dateUtils';

export type DailyTotal = {
  date: string;
  minutes: number;
};

export type WeeklyTotal = {
  weekStart: string;
  minutes: number;
  sessionCount: number;
};

export type SubjectBreakdown = {
  name: string;
  minutes: number;
  sessionCount: number;
  percentage: number;
};

export type HourlyDistribution = {
  hour: number;
  minutes: number;
};

export type MoodDistribution = {
  mood: string;
  count: number;
  percentage: number;
};

/**
 * Compute daily totals for a date range.
 */
export function computeDailyTotals(
  sessions: StudySession[],
  startDate: Date,
  endDate: Date
): DailyTotal[] {
  const map = new Map<string, number>();
  const dates = dateRange(startDate, endDate);

  for (const d of dates) {
    map.set(d.toISOString().split('T')[0], 0);
  }

  for (const session of sessions) {
    const dateKey = session.startAt.split('T')[0];
    if (map.has(dateKey)) {
      map.set(dateKey, (map.get(dateKey) ?? 0) + session.durationMinutes);
    }
  }

  return Array.from(map.entries())
    .map(([date, minutes]) => ({ date, minutes }))
    .sort((a, b) => a.date.localeCompare(b.date));
}

/**
 * Compute weekly totals for a given number of weeks back.
 */
export function computeWeeklyTotals(
  sessions: StudySession[],
  weeksBack: number = 12
): WeeklyTotal[] {
  const now = new Date();
  const weeks: WeeklyTotal[] = [];

  for (let i = weeksBack - 1; i >= 0; i--) {
    const weekStart = startOfWeek(new Date(now.getTime() - i * 7 * 24 * 60 * 60 * 1000));
    const weekEnd = new Date(weekStart.getTime() + 7 * 24 * 60 * 60 * 1000 - 1);

    const weekSessions = sessions.filter((s) => {
      const d = new Date(s.startAt);
      return d >= weekStart && d <= weekEnd;
    });

    weeks.push({
      weekStart: weekStart.toISOString().split('T')[0],
      minutes: weekSessions.reduce((sum, s) => sum + s.durationMinutes, 0),
      sessionCount: weekSessions.length
    });
  }

  return weeks;
}

/**
 * Compute subject breakdown with percentages.
 */
export function computeSubjectBreakdown(sessions: StudySession[]): SubjectBreakdown[] {
  const map = new Map<string, { minutes: number; count: number }>();
  let totalMinutes = 0;

  for (const session of sessions) {
    const name = session.subjectName ?? 'Uncategorized';
    const existing = map.get(name) ?? { minutes: 0, count: 0 };
    existing.minutes += session.durationMinutes;
    existing.count += 1;
    map.set(name, existing);
    totalMinutes += session.durationMinutes;
  }

  return Array.from(map.entries())
    .map(([name, data]) => ({
      name,
      minutes: data.minutes,
      sessionCount: data.count,
      percentage: totalMinutes > 0 ? Math.round((data.minutes / totalMinutes) * 100) : 0
    }))
    .sort((a, b) => b.minutes - a.minutes);
}

/**
 * Compute hourly distribution of study activity.
 */
export function computeHourlyDistribution(sessions: StudySession[]): HourlyDistribution[] {
  const hours: number[] = new Array(24).fill(0);

  for (const session of sessions) {
    const hour = getHour(session.startAt);
    hours[hour] += session.durationMinutes;
  }

  return hours.map((minutes, hour) => ({ hour, minutes }));
}

/**
 * Find peak study hour.
 */
export function findPeakHour(sessions: StudySession[]): number {
  const distribution = computeHourlyDistribution(sessions);
  let peakHour = 0;
  let maxMinutes = 0;

  for (const { hour, minutes } of distribution) {
    if (minutes > maxMinutes) {
      maxMinutes = minutes;
      peakHour = hour;
    }
  }

  return peakHour;
}

/**
 * Compute mood distribution.
 */
export function computeMoodDistribution(sessions: StudySession[]): MoodDistribution[] {
  const map = new Map<string, number>();
  let total = 0;

  for (const session of sessions) {
    if (session.mood) {
      map.set(session.mood, (map.get(session.mood) ?? 0) + 1);
      total++;
    }
  }

  return Array.from(map.entries())
    .map(([mood, count]) => ({
      mood,
      count,
      percentage: total > 0 ? Math.round((count / total) * 100) : 0
    }))
    .sort((a, b) => b.count - a.count);
}

/**
 * Compute rolling average of daily study time.
 */
export function computeRollingAverage(dailyTotals: DailyTotal[], windowDays: number): number {
  if (dailyTotals.length === 0) return 0;

  const recent = dailyTotals.slice(-windowDays);
  const totalMinutes = recent.reduce((sum, d) => sum + d.minutes, 0);
  return Math.round(totalMinutes / recent.length);
}

/**
 * Compute study/break ratio.
 */
export function computeStudyBreakRatio(sessions: StudySession[]): {
  studyMinutes: number;
  breakMinutes: number;
  ratio: string;
} {
  const studyMinutes = sessions.reduce((sum, s) => sum + s.durationMinutes, 0);
  const breakMinutes = sessions.reduce((sum, s) => sum + (s.breakMinutes ?? 0), 0);

  const ratio =
    breakMinutes > 0
      ? `${(studyMinutes / breakMinutes).toFixed(1)}:1`
      : studyMinutes > 0
        ? '∞:1'
        : '0:0';

  return { studyMinutes, breakMinutes, ratio };
}
